### PR TITLE
chore: use sentinel proofs for localhost channels

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Creates a sentinel proof used for localhost channels. 
- Re-add validation of proof lengths in 04-channel msg validate basic methods

closes: #3194 

### Commit Message / Changelog Entry

**NA**

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/light-clients/09-localhost/keys.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package localhost
+
+const (
+	// ModuleName defines the 09-localhost light client module name
+	ModuleName = "09-localhost"
+)
+
+// SentinelProof defines the 09-localhost sentintel proof
+var SentinelProof = []byte("proof_localhost")
```
Comment (colin-axner) [inline modules/light-clients/09-localhost/keys.go:9]: I think we should explain a little of what `sentinel proof` means?

**Review by colin-axner — APPROVED** [review]
LGTM! I left suggestions on doc improvements/error msgs

**Review by chatton — APPROVED** [review]
Looks great, I love how simple this actually was to implement! 

I agree with @colin-axner's comments about making it a clear why we are doing the comparison in the first place on the sentinel value.

File: modules/core/04-channel/types/msgs.go [hunk 2]
```diff
@@ -91,6 +92,9 @@ func (msg MsgChannelOpenTry) ValidateBasic() error {
 	if msg.PreviousChannelId != "" {
 		return sdkerrors.Wrap(ErrInvalidChannelIdentifier, "previous channel identifier must be empty, this field has been deprecated as crossing hellos are no longer supported")
 	}
+	if len(msg.ProofInit) == 0 {
+		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "cannot submit an empty proof init")
```
Comment (crodriguezvega) [inline modules/core/04-channel/types/msgs.go:96]:
```suggestion
		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "cannot submit an empty init proof")
```

File: modules/core/04-channel/types/msgs.go [hunk 3]
```diff
@@ -149,6 +153,9 @@ func (msg MsgChannelOpenAck) ValidateBasic() error {
 	if err := host.ChannelIdentifierValidator(msg.CounterpartyChannelId); err != nil {
 		return sdkerrors.Wrap(err, "invalid counterparty channel ID")
 	}
+	if len(msg.ProofTry) == 0 {
+		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "cannot submit an empty proof try")
```
Comment (crodriguezvega) [inline modules/core/04-channel/types/msgs.go:157]:
```suggestion
		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "cannot submit an empty try proof")
```

File: modules/core/04-channel/types/msgs.go [hunk 4]
```diff
@@ -191,6 +198,9 @@ func (msg MsgChannelOpenConfirm) ValidateBasic() error {
 	if !IsValidChannelID(msg.ChannelId) {
 		return ErrInvalidChannelIdentifier
 	}
+	if len(msg.ProofAck) == 0 {
+		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "cannot submit an empty proof ack")
```
Comment (crodriguezvega) [inline modules/core/04-channel/types/msgs.go:202]:
```suggestion
		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "cannot submit an empty acknowledgement proof")
```

File: modules/core/04-channel/types/msgs.go [hunk 5]
```diff
@@ -272,6 +282,9 @@ func (msg MsgChannelCloseConfirm) ValidateBasic() error {
 	if !IsValidChannelID(msg.ChannelId) {
 		return ErrInvalidChannelIdentifier
 	}
+	if len(msg.ProofInit) == 0 {
+		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "cannot submit an empty proof init")
```
Comment (crodriguezvega) [inline modules/core/04-channel/types/msgs.go:286]:
```suggestion
		return sdkerrors.Wrap(commitmenttypes.ErrInvalidProof, "cannot submit an empty init proof")
```

File: modules/light-clients/09-localhost/keys.go [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package localhost
+
+const (
+	// ModuleName defines the 09-localhost light client module name
+	ModuleName = "09-localhost"
+)
+
+// SentinelProof defines the 09-localhost sentintel proof
```
Comment (crodriguezvega) [inline modules/light-clients/09-localhost/keys.go:8]:
```suggestion
// SentinelProof defines the 09-localhost sentinel proof
```

File: modules/light-clients/09-localhost/keys.go (hunk 1 above)
Comment (crodriguezvega) [reply modules/light-clients/09-localhost/keys.go:9]: I am not sure, but I think I remember @AdityaSripal saying that we could just use a slice with one byte?

File: modules/core/04-channel/types/msgs_test.go [hunk 7]
```diff
@@ -55,6 +55,8 @@ var (
 	packet        = types.NewPacket(validPacketData, 1, portid, chanid, cpportid, cpchanid, timeoutHeight, timeoutTimestamp)
 	invalidPacket = types.NewPacket(unknownPacketData, 0, portid, chanid, cpportid, cpchanid, timeoutHeight, timeoutTimestamp)
 
+	emptyProof = []byte{}
```
Comment (crodriguezvega) [inline modules/core/04-channel/types/msgs_test.go:58]: Would it make sense to also have tests for a nil proof?

**Review by crodriguezvega — COMMENTED** [review]
Nice idea, @damiannolan!

File: modules/core/04-channel/types/msgs_test.go (hunk 7 above)
Comment (damiannolan) [reply modules/core/04-channel/types/msgs_test.go:58]: I think its probably overkill. We check `len(proof) != 0` and the built-in `len` function treats `[]byte{}` and `nil` the same.

File: modules/light-clients/09-localhost/keys.go (hunk 1 above)
Comment (damiannolan) [reply modules/light-clients/09-localhost/keys.go]: Updated
